go-controller/pkg/node/port_claim.go

```diff
@@ -0,0 +1,202 @@
+package node
+
+import (
+	"fmt"
+	"net"
+	"reflect"
+	"sync"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+	kapi "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog"
+)
+
+var port localPort
+
+type activeSocket interface {
+	Close() error
+}
+
+type localPort interface {
+	open(port int32, protocol kapi.Protocol, svc *kapi.Service) error
+	close(port int32, svc *kapi.Service) error
+}
+
+type portClaimWatcher struct {
+	recorder          record.EventRecorder
+	activeSocketsLock sync.Mutex
+	activeSockets     map[int32]activeSocket
```

Comment:
L31: How will this work if we have a service with external IP such that it has same port number of TCP and UDP, as in,

```
spec:
  clusterIP: 10.223.0.3
  ports:
  - name: dns
    port: 53
    protocol: UDP
    targetPort: 53
  - name: dns-tcp
    port: 53
    protocol: TCP
    targetPort: 53
  - name: metrics
    port: 9153
    protocol: TCP
    targetPort: 9153
```